internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,308 @@
+	if req == nil || req.Spec.Target.Kind == constant.EmptyString {
+		return ctrl.Result{}, nil
+	}
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling NonAdminDownloadRequest", "name", req.Name, "namespace", req.Namespace)
+	// defines associated downloadrequest for getting, or deleting
+	veleroDR := velerov1.DownloadRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.VeleroDownloadRequestName(),
+			Namespace: r.OADPNamespace,
+			Labels: func() map[string]string {
+				nal := function.GetNonAdminLabels()
+				nal[constant.NadrOriginNACUUIDLabel] = string(req.GetUID())
+				return nal
+			}(),
+			Annotations: function.GetNonAdminDownloadRequestAnnotations(req),
+		},
+		Spec: velerov1.DownloadRequestSpec{
+			Target: velerov1.DownloadTarget{
+				Kind: req.Spec.Target.Kind,
+			},
+		},
+	}
+	// if request has status, it may already be processed or expired
+	if req.Status.VeleroDownloadRequest.Status != nil {
+		// if request is expired, delete NADR after deleting velero DR
+		if req.Status.VeleroDownloadRequest.Status.Expiration != nil && req.Status.VeleroDownloadRequest.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+			// find associated velero downloadrequest and delete that first
+			return reconcile.Result{}, r.deleteExpiredDownloadRequest(ctx, &veleroDR, req)
+		}
+		// if request is not expired, and has downloadUrl, requeue when expired to cleanup.
+		if req.Status.VeleroDownloadRequest.Status.DownloadURL != constant.EmptyString {
+			return reconcile.Result{RequeueAfter: time.Until(req.Status.VeleroDownloadRequest.Status.Expiration.Time)}, nil
+		}
+	}
+	// try get veleroDR if exists, then update status
+	if err := r.Get(ctx, types.NamespacedName{Namespace: veleroDR.Namespace, Name: veleroDR.Name}, &veleroDR); err == nil {
+		return reconcile.Result{}, r.updateNADRWithURL(ctx, &veleroDR, req)
+	} else if !apierrors.IsNotFound(err) {
+		// some other errors, requeue to retry get
+		return reconcile.Result{}, err
+	}
+	// veleroDR is not found, so we create one
+	return reconcile.Result{}, r.processDownloadRequest(ctx, &veleroDR, req)
+}
+
+// Delete expired download request
+// find associated velero downloadrequest and delete that first
+func (r *NonAdminDownloadRequestReconciler) deleteExpiredDownloadRequest(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("Deleting expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), req.Namespace)
+	if err := r.Delete(ctx, veleroDR); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), err)
+			// other errors, requeue to retry delete
+			return err
+		}
+	}
+	logger.V(1).Info("Deleting expired NonAdminDownloadRequest", req.Name, req.Namespace)
```

Comment:
to keep similar behavior for backup/restore, should we just add message in status that associated velero DownloadRequest was deleted?

Reply:
DownloadRequest life is very short lived.. so I don't think knowing velero is deleted is useful information to patch into the object. we already logged velero object is being deleted.

Reply:
could patch in.. but the object is being deleted right after.. I don't think you'd want this object lingering past expiration IMO.